torch/_higher_order_ops/associative_scan.py

```diff
@@ -0,0 +1,173 @@
+            i.e. ``combine_fn(a, combine_fn(b, c)) == combine_fn(combine_fn(a, b), c)``
+
+    Example::
+
+        def add(x: torch.Tensor, y: torch.Tensor):
+            return x + y
+
+        cumsum = associative_scan(x, dim, combine_fn=add)
+
+    """
+
+    torch._check(isinstance(input, torch.Tensor), "input must be a Tensor")
+    dim = utils.canonicalize_dim(input.ndim, dim)
+    torch._check(callable(combine_fn), "combine_fn must be a callable")
+
+    if torch._dynamo.is_compiling():
+        return associative_scan_op(input, dim, combine_fn)
+
+    if not torch._dynamo.is_dynamo_supported():
+        raise RuntimeError("associative_scan requires dynamo support.")
+
+    with _set_compilation_env(), torch._dynamo.utils.disable_cache_limit():
+        return torch.compile(associative_scan_op, fullgraph=True)(
+            input, dim, combine_fn
+        )
+
+
+associative_scan_op = HigherOrderOperator("associative_scan")
+
+
+def trace_associative_scan(proxy_mode, func_overload, input, dim, combine_fn):
+    pre_dispatch = getattr(proxy_mode, "pre_dispatch", False)
+
+    with disable_proxy_modes_tracing():
+        sample_inputs = (
+            torch.full((), False, dtype=input.dtype, device=input.device),
+            torch.full((), False, dtype=input.dtype, device=input.device),
+        )
+        combine_graph = make_fx(
+            _maybe_run_with_interpreter(combine_fn), pre_dispatch=pre_dispatch
+        )(*sample_inputs)
+
+    outputs = []
+    for node in combine_graph.graph.nodes:
+        if node.op == "output":
+            outputs.extend(node.args)
+
+    torch._check(
+        len(outputs) == 1,
+        lambda: f"expected combine_fn to have 1 output but got {len(outputs)}",
+    )
+
+    for o in outputs:
+        o_meta = o.meta["tensor_meta"]
+        torch._check(
+            o_meta.dtype == input.dtype,
+            lambda: (
+                f"combine_fn output type mismatch, expected {input.dtype} "
+                + f"but got {o_meta.dtype}"
+            ),
+        )
+        torch._check(
+            o_meta.shape == (),
+            lambda: f"combine_fn must return a scalar tensor but got shape {o_meta.shape}",
+        )
+
+    _, combine_graph_name = unique_graph_id(proxy_mode, prefix="scan_combine_graph")
+
+    proxy_mode.tracer.root.register_module(combine_graph_name, combine_graph)
+
+    args = (input, dim, combine_graph)
+    proxy_args = pytree.tree_map(proxy_mode.tracer.unwrap_proxy, args)
+    out_proxy = proxy_mode.tracer.create_proxy(
+        "call_function", func_overload, proxy_args, {}, name="associative_scan"
+    )
+
+    with disable_proxy_modes_tracing():
+        out = aten.clone(input)
+
+    return track_tensor_tree(out, out_proxy, constant=None, tracer=proxy_mode.tracer)
+
+
+@associative_scan_op.py_impl(DispatchKey.CompositeExplicitAutograd)
+def associative_scan_op_dense(input, dim, combine_fn):
+    raise NotImplementedError("associative_scan is not implemented for eager")
```

Comment:
Wondering if this op fundamentally can't be implemented in eager or this is temporary? In principle, we could fallback to a serialized version in terms of eager ops?

Reply:
Yes there's no reason we couldn't implement this in eager with pointwise operations and shuffling the tensor, it just wouldn't perform as well.